### ISSUE TITLE
Any plan on 2D game support?

### ISSUE THREAD

**Issue by Jack-Ji** [michal-z/zig-gamedev#95]
Hi @michal-z , I've being following your work closely, nice work!

A quick question, will you consider add libs for 2d games? e.g. box2d/vector graphics/spine animation etc.

I'm asking because I think most indie developers choose 2d over 3d game (at least initially), since a decent 3d game requires way more expertise and amount of work.

**Comment by michal-z** [comment]
Hi, yes I would like to support 2D gamedev. We can use this issue for feature requests/discussion.

Regarding physics, I think we could use Jolt also for 2D. I will investigate this and ask the author of the lib about perf implications.

**Comment by Jack-Ji** [comment]
Great to heart that! Here's my wishlist:

* Sprites batching support (with optional shader effect)
* Tiled editor support
* 2D Physics
* Vector graphics
* Font system
* Animation system (3d also)
* Particle system (3d also)

Might be too much, feel free to take your pick 😊😁.

**Comment by michal-z** [comment]
Regarding font system and UI - for my game, I plan to use font and UI system from dear imgui (zgui).

**Comment by Jack-Ji** [comment]
Hmm, never thought of this. I only use imgui to make tools. For a game, I might choose
a more stylish one (with some animation behavior and colorful skins). Not sure if imgui can
deliver that.

**Comment by michal-z** [comment]
I think it can with some code tweaks. I will be trying to do it for my games because I don't want to have another UI/font system. Will see how it goes.

**Comment by Pyrolistical** [comment]
I know this is exactly what you want, but I've implemented a 2D sample https://github.com/michal-z/zig-gamedev/tree/main/samples/instanced_pills_wgpu

What might be of interest to you is the sample does an object space to clip space transform with correct aspect ratio. https://github.com/michal-z/zig-gamedev/blob/main/samples/instanced_pills_wgpu/src/InstancedPillsWgpu.zig#L629

I'm only interested in 2D as well and will simply use WebGPU with z == 0.

**Comment by michal-z** [comment]
@Pyrolistical Looks nice. I think we can make a sample from it and then tweak, improve and potentially make a lib from this.

**Comment by Pyrolistical** [comment]
@michal-z the font system in imgui seems pretty limited. ootb it doesn't support rotating/transform/scaling of text

**Comment by michal-z** [comment]
Do you mean our `zgui` wrapper or the `dear imgui` lib?

**Comment by Pyrolistical** [comment]
Both? Unless you've added this functionality to zgui?

**Comment by michal-z** [comment]
Nvm, I thought you mean that this functionality is present in `dear imgui` but that we just don't expose it yet.

**Comment by michal-z** [comment]
We could render imgui text to a texture and then apply a transformation when drawing this texture to the back buffer.

**Comment by Pyrolistical** [comment]
Yeah, that was my first thought as well, but the texture would need to be oversized in order to compensate for the rotation blur. Doable but cumbersome.

The ideal solution would transformed the text in the shader and render out pixel perfect.

**Comment by michal-z** [comment]
Yes, trying it out is good idea, thanks!

**Comment by michal-z** [comment]
Jolt Physics should get 2D support soon.

**Comment by Jack-Ji** [comment]
> Jolt Physics should get 2D support soon.

How's things going with 2d support of zphysics? No rush, just checking...

**Comment by michal-z** [comment]
C++ version already supports 2D. Should be pretty easy to have it in zphysics. Will research and get back to you next weak.

**Comment by michal-z** [comment]
Jolt supports 2D by suing constraints and locking one axis. There is a 2D example in the C++ added in this commit:
https://github.com/jrouwe/JoltPhysics/commit/4f979680567d4de1a7bf7d4eb7302307ea892c8e

To get this functionality we need to add "constraints" to our bindings. I will start working on it on Friday I think.

**Comment by michal-z** [comment]
Actually it seems that there is a 2D branch:
https://github.com/jrouwe/JoltPhysics/tree/feature/2d_sim

**Comment by michal-z** [comment]
I'm considering creating bindings for this lib as a 2D solution for zig-gamedev: https://github.com/grimfang4/sdl-gpu

What do you think about this?

**Comment by Jack-Ji** [comment]
Frankly, with SDL3's gpu api being actively worked, I wouldn't be very interested in sdl_gpu. Keep improving zsdl is enough to my opinion.

**Comment by michal-z** [comment]
> Frankly, with SDL3's gpu api being actively worked, I wouldn't be very interested in sdl_gpu. Keep improving zsdl is enough to my opinion.

Is SDL3 gpu API work done in public? I've only seen a very rough header file. My concern is that it may take years to come (making a shader language is a huge rabbit hole).

Also, I have a feeling that official SDL3 GPU API will be a lower level HW abstraction.

But yeah, I get your point. Thanks!

**Comment by michal-z** [comment]
Here is a link to SDL3 GPU API PR: https://github.com/libsdl-org/SDL/pull/7067

**Comment by michal-z** [comment]
Looks like a minimal HW abstraction layer (no batching, etc.) Looks nice but I think that something like sdl_gpu will still be needed for 2D games. Thoughts?

**Comment by michal-z** [comment]
Note that sdl_gpu builds on top of SDL2. It basically provides batching and better sprite support to SDL2.

**Comment by Jack-Ji** [comment]
Sorry, didn't realize that. In that case, more ziggy API for sdl-gpu does sound tempting indeed.
